overwatch/src/services/relay.rs

```diff
@@ -21,20 +23,52 @@ pub enum RelayError {
     Receiver(Box<dyn Debug + Send + Sync>),
 }
 
+#[derive(Error, Debug)]
+pub enum ServiceError {
+    #[error("Couldn't start service")]
+    Start,
+    #[error("Couldn't stop service")]
+    Stop,
+}
+
 /// Message wrapper type.
 pub type AnyMessage = Box<dyn Any + Send + 'static>;
 
-/// Result type when creating a relay connection.
-pub type RelayResult = Result<AnyMessage, RelayError>;
+/// Channel to retrieve the consumer of the relay connection.
+/// The intended usage is oneshot-like, but having them as mpsc simplifies
+/// reusing the relay when a service is stopped and started.
+pub type ConsumerSender<Message> = sync_mpsc::Sender<Receiver<Message>>;
+pub type ConsumerReceiver<Message> = sync_mpsc::Receiver<Receiver<Message>>;
 
 /// Channel receiver of a relay connection.
 #[derive(Debug)]
 pub struct InboundRelay<Message> {
     receiver: Receiver<Message>,
+    /// Sender to return the consumer back to the caller
+    /// This is used to maintain a single consumer while being able to reuse it
+    /// when the same service is stopped and started.
+    consumer_sender: ConsumerSender<Message>,
```

Comment:
Why do we need to reuse it? Can't overwatch generate a new one upon a new service start?

Reply:
I don't think so, because otherwise we would need to reset the senders on other services. I think its simpler this way.

Reply:
I though would always intermediate communications? Isn't this only used by overwatch to communicate with the service?

Reply:
> I though would always intermediate communications? Isn't this only used by overwatch to communicate with the service?

No, overwatch just kinda plugs cables together, it doesnt do anything itself. So when a service is gone, it just holds one end of the cable until it have to give it to the new instance. We could go fancier probably, but I'm not sure it would be worth it.

Reply:
Mhh yeah ok I saw now inside the `ServiceRunner`. I guess I would have probably gone a different route, like creating new channels when spawning a new service, with the senders stored within overwatch and the receivers given to the runner. The sender could then be cloned and given to other parties for, e.g., subscriptions. But that requires more changes, so probably not suitable now. In this case, I will resolve this comment, since that answers my question.

Reply:
> Mhh yeah ok I saw now inside the `ServiceRunner`. I guess I would have probably gone a different route, like creating new channels when spawning a new service, with the senders stored within overwatch and the receivers given to the runner. The sender could then be cloned and given to other parties for, e.g., subscriptions. But that requires more changes, so probably not suitable now. In this case, I will resolve this comment, since that answers my question.

Yeah, but the problem of that would be that you need to refresh any sender that some other service may hold already. I thought about that as well, but we would have to kinda wrap it and make a refresh method to request and updated sender or something like that. So, TLDR, fair but probably not necessary.

Reply:
But why? The sender would be created once by overwatch and sent to all services. That would not change as overwatch would hold the same receiver until it's shut down. What changes is a _second_ sender instance that it holds with each service separately, that those services can use to communicate with each other via overwatch.

Reply:
So one sender out to everyone and overwatch holds the receiver, and one receiver for each service where overwatch holds the sender.

Reply:
> So one sender out to everyone and overwatch holds the receiver, and one receiver for each service where overwatch holds the sender.

Ok, so you mean Overwatch to be a man in the middle to relay from ones to the others. It could be, we would have to spawn a task for that probably (unless there is something that abstracts it available already 🤔 ). In any case, its the same effect as it is right now. And I kinda prefer not to have an extra task if possible.

Reply:
I tried not to deviate from the existing architecture. We can always do modifications later on :)